x-pack/plugins/canvas/canvas_plugin_src/plugin.ts

```diff
@@ -37,9 +40,10 @@ export class CanvasSrcPlugin implements Plugin<void, void, SetupDeps, StartDeps>
 
     plugins.canvas.addRenderers(renderFunctions);
 
-    core.getStartServices().then(([coreStart, depsStart]) => {
+    core.getStartServices().then(([coreStart, startPlugins]) => {
```

Comment:
this doesn't look right. you await `core.getStartServices()` here which will only resolve once the start lifectcle completes. this means that your setup lifecycle actually completes after start.

what you should rather do is pass `core.getStartServices` method to your factory and await on it inside  your function body. this way the setup lifecycle will be done in time as we'll only await on start services when running the function (which can only happen after start lifecycle)

also its not clear to me if this functions are reistered on both client and server (they should be)

take a look at how data plugin does it for example:
[here](https://github.com/elastic/kibana/blob/master/src/plugins/data/public/search/search_service.ts#L121) we are registering a function and passing `core.getStartServices to the factory`
[here](https://github.com/elastic/kibana/blob/master/src/plugins/data/public/search/expressions/kibana_context.ts) and [here](https://github.com/elastic/kibana/blob/master/src/plugins/data/server/search/expressions/kibana_context.ts) you can see how we wrap core services so we can pass a common thing to the function on client and server. note that this is still just async function being passed to actual kibana_context factory (so nothing executed/awaited for at setup time)
[here](https://github.com/elastic/kibana/blob/master/src/plugins/data/common/search/expressions/kibana_context.ts#L124) you can see that how we await on core.getStartServices inside function body of kibana_context

Reply:
So I've been looking this over, and this feedback is entirely valid and really important, but unfortunately I don't think it can be addressed in this PR.  We will most certainly address it in #105675.

@ppisljar My choices are to commit as-is, or fix this one case the correct way and leave the others.  How do you feel about me committing this to match the current (admittedly incorrect) pattern and fix all of this in #105675?  Otherwise this legacy service is going to have to stick around until we address #105675.

Reply:
I am fine leaving this as it is in this PR, we are not changing it for worse and if you think you are gonna reach the end goal easier in steps that is completely ok with me.